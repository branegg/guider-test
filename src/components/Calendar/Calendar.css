.Calendar {
  width: 100%;
  height: 100%;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  margin: 50px 0;
}

.Calendar__header {
  font-size: 18px;
  font-weight: 500;
  margin-bottom: 30px;
}

.Calendar__wrapper {
  display: flex;
  justify-content: center;
  align-items: center;
}

.Calendar__item {
  width: 120px;
  height: 80px;
  border-radius: 5px;
  border: 2px solid #f8f9fa;
  background-color: #f8f9fa;
  transition: all 0.15s ease-in-out;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  color: #c2c4c6;
  margin: 0 2px;
}

.Calendar__item--available {
  color: black;
  border: 2px solid #eceef2;
  background-color: #eceef2;
}

.Calendar__item--available:hover {
  border: 2px solid rgba(90, 92, 234, 0.3);
  background-color: rgba(90, 92, 234, 0.3);
  cursor: pointer;
}

.Calendar__item--active {
  color: #5a5cea;
  border: 2px solid #5a5cea;
  background-color: #f8f9fa;
}

.Calendar__item--active .Calendar__date {
  font-weight: 900;
}

.Calendar__item--active .Calendar__date--small {
  font-weight: 600;
  margin-left: -1px;
}

.Calendar__item--active .Calendar__day {
  font-weight: 900;
}

.Calendar__item--active:hover {
  border: 2px solid #5a5cea;
  background-color: #f8f9fa;
  cursor: pointer;
}

.Calendar__date {
  font-size: 20px;
  font-weight: 700;
  margin-bottom: 10px;
  transition: all 0.15s ease-in-out;
}

.Calendar__date--small {
  font-size: 14px;
  font-weight: 500;
  margin-left: -2px;
  transition: all 0.15s ease-in-out;
}

.Calendar__day {
  font-weight: 700;
  font-size: 20px;
  transition: all 0.15s ease-in-out;
}
