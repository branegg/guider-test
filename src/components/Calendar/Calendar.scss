@import './../../constants/colors.scss';

.Calendar {
  width: 100%;
  height: 100%;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  margin: 50px 0;

  &__header {
    font-size: 18px;
    font-weight: 500;
    margin-bottom: 30px;
  }
  &__wrapper {
    display: flex;
    justify-content: center;
    align-items: center;
  }
  &__item {
    width: 120px;
    height: 80px;
    border-radius: 5px;
    border: 2px solid $lightGray;
    background-color: $lightGray;
    transition: all 0.15s ease-in-out;
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    color: $fontGray;
    margin: 0 2px;

    &--available {
      color: black;
      border: 2px solid $gray;
      background-color: $gray;

      &:hover {
        border: 2px solid rgba($purple, 0.3);
        background-color: rgba($purple, 0.3);
        cursor: pointer;
      }
    }

    &--active {
      color: $purple;
      border: 2px solid $purple;
      background-color: $lightGray;
      .Calendar__date {
        font-weight: 900;

        &--small {
          font-weight: 600;
          margin-left: -1px;
        }
      }
      .Calendar__day {
        font-weight: 900;
      }

      &:hover {
        border: 2px solid $purple;
        background-color: $lightGray;
        cursor: pointer;
      }
    }
  }
  &__date {
    font-size: 20px;
    font-weight: 700;
    margin-bottom: 10px;
    transition: all 0.15s ease-in-out;
    &--small {
      font-size: 14px;
      font-weight: 500;
      margin-left: -2px;
      transition: all 0.15s ease-in-out;
    }
  }
  &__day {
    font-weight: 700;
    font-size: 20px;
    transition: all 0.15s ease-in-out;
  }
}
