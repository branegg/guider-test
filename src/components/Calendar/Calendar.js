import './Calendar.scss';
import React, { useEffect, useState } from 'react';
import { DateTime } from 'luxon';
import CalendarItem from './CalendarItem/CalendarItem';
import classNames from 'classnames';
import axios from 'axios';

import loadingSpinner from './../../assets/images/loading.gif';

const Calendar = () => {
  const [week, setWeek] = useState([]);
  const [availability, setAvailability] = useState([]);
  const [activeDate, setActiveDate] = useState();
  const [hours, setHours] = useState([]);
  const [availableHours, setAvailableHours] = useState([]);
  const [isLoading, setIsLoading] = useState(true);

  const availableDates = availability && availability.map((date) => date.date);

  useEffect(() => {
    prepareWeek();
    prepareHours();
    fetchAvailability();
  }, []);

  const fetchAvailability = async () => {
    const response = await axios.get('/availability');
    const data = response.data;
    setAvailability(data);
    setIsLoading(false);
  };

  const prepareWeek = () => {
    let currentWeek = [];
    for (let i = 0; i <= 6; i++) {
      currentWeek.push(DateTime.local().startOf('week').plus({ days: i }));
    }
    setWeek(currentWeek);
  };

  const prepareHours = () => {
    const dayStart = '08:00';
    const dayEnd = '17:00';
    let hours = [];

    for (
      let i = dayStart;
      i !== dayEnd;
      i = DateTime.fromISO(i).plus({ hours: 1 }).toLocaleString(DateTime.TIME_24_SIMPLE)
    ) {
      hours.push(i);
    }

    setHours(hours);
  };

  const renderDays = () => {
    return week.map((day) => {
      const isAvailable = availableDates.includes(day.toFormat('dd/MM/yyyy'));
      return (
        <CalendarItem
          key={day.ts}
          dayObject={day}
          isAvailable={isAvailable}
          onClick={() => isAvailable && handleDateClick(day.ts)}
          isActive={activeDate === day.ts}
        />
      );
    });
  };

  const renderHours = () =>
    hours.map((hour, index) => (
      <p
        data-testid='calendar-hour'
        key={index}
        className={classNames({ Calendar__hour: true, 'Calendar__hour--active': availableHours.includes(hour) })}
      >
        {hour}
      </p>
    ));

  const renderLoadingSpinner = () => (
    <div className='Calendar__loading'>
      <img className='Calendar__spinner' src={loadingSpinner} alt='Loading...' />
    </div>
  );

  const handleDateClick = (key) => {
    const activeDate = week.filter((day) => day.ts === key)[0].toFormat('dd/MM/yyyy');
    setActiveDate(key);
    const availableSlots = availability.filter((date) => date.date === activeDate)[0].availableSlots;
    setAvailableHours(availableSlots.map((slot) => slot.startTime));
  };

  return (
    <div className='Calendar'>
      {isLoading && renderLoadingSpinner()}
      <h2 className='Calendar__header'>Your timezone: {DateTime.local().zoneName}</h2>
      <section className='Calendar__wrapper' data-testid='calendar-wrapper'>
        {renderDays()}
      </section>
      <section className='Calendar__hours'>{renderHours()}</section>
    </div>
  );
};

export default Calendar;
