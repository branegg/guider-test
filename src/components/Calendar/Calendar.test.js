import React from 'react';
import Calendar from './Calendar';
import { render, cleanup, screen, act, fireEvent } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { DateTime } from 'luxon';
import axiosMock from 'axios';

jest.mock('axios');

const mockResponse = [
  {
    date: DateTime.local().toFormat('dd/MM/yyyy'),
    availableSlots: [
      { startTime: '09:00', endTime: '10:00' },
      { startTime: '10:00', endTime: '11:00' },
    ],
  },
  {
    date: DateTime.local().plus({ days: 1 }).toFormat('dd/MM/yyyy'),
    availableSlots: [
      { startTime: '13:00', endTime: '14:00' },
      { startTime: '15:00', endTime: '16:00' },
      { startTime: '16:00', endTime: '17:00' },
    ],
  },
];

beforeEach(() => {
  axiosMock.get.mockResolvedValueOnce({
    data: mockResponse,
  });
});

afterEach(cleanup);

describe('Calendar', () => {
  it('renders without crashing', async () => {
    await act(async () => render(<Calendar />));
  });

  it('matches snapshot', async () => {
    const tree = await renderer.act(async () => await renderer.create(<Calendar />).toJSON());
    await expect(tree).toMatchSnapshot();
  });

  it('renders full week', async () => {
    await act(async () => render(<Calendar />));
    const calendarWrapper = screen.getByTestId('calendar-wrapper');
    expect(calendarWrapper.children.length).toBe(7);
  });

  it('renders available dates', async () => {
    await act(async () => render(<Calendar />));
    const availableDates = screen
      .getAllByTestId('calendar-item')
      .filter((item) => item.classList.contains('Calendar__item--available'));

    expect(availableDates.length).toBe(2);
  });

  it('handles date clicks', async () => {
    await act(async () => render(<Calendar />));
    const availableDates = screen
      .getAllByTestId('calendar-item')
      .filter((item) => item.classList.contains('Calendar__item--available'));

    const checkActiveHours = () => {
      const calendarHours = screen.getAllByTestId('calendar-hour');
      const activeHours = calendarHours.filter((hour) => hour.classList.contains('Calendar__hour--active'));
      return activeHours.length;
    };

    expect(checkActiveHours()).toBe(0);

    fireEvent.click(availableDates[0]);

    const activeDate = screen
      .getAllByTestId('calendar-item')
      .filter((item) => item.classList.contains('Calendar__item--active'));

    expect(activeDate.length).toBe(1);
    expect(checkActiveHours()).toBe(2);

    fireEvent.click(availableDates[1]);

    expect(activeDate.length).toBe(1);
    expect(checkActiveHours()).toBe(3);
  });
});
