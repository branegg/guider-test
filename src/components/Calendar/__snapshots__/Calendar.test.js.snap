// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calendar matches snapshot 1`] = `undefined`;
